azure-iot-device/azure/iot/device/patch.py

```diff
@@ -0,0 +1,133 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""This module provides patches used to dynamically modify items from the libraries"""
```

Comment:
damn, this is a really complex piece of code, just to make sure that class methods show up in the documentation as a class method instead of an instance method.  I've been reading this for about an hour now trying to find ways to make it simpler.  I've learned a ton about how name binding and closures and exec all work.  I haven't found anything to make it simpler, but I am finding that every question I have leads down a rabbit-hole.  I suspect this will be a "carter is the only one who can touch this code without breaking it" module, and that scares me.

Reply:
One thing that I think would help in maintainability would be to add debug output for every exec call that lets developers grep and read the debug output almost like it's a source file.  e.g.:
```
INFO:azure.iot.device.iothub.patch: exec: from azure.iot.device import IoTHubModuleClient as _IoTHubModuleClient
INFO:azure.iot.device.iothub.patch: exec: def receive_input_message(self, input_name, block=True, timeout=None): return super(_IotHubModuleClient, self).receive_input_message(self=self, input_name=input_name, block=block, timeout=timeout))
INFO:azure.iot.device.iothub.patch: exec: receive_input_message.__doc__ = method_object.__doc__
INFO:azure.iot.device.iothub.patch: exec: setattr(target_class, method_name, receive_input_message)
```

This makes me think you conceptionally have two namespaces here, both living inside globals():
1. The namspace that you get from being a python module without using exec().  Things like method_object and target_class fall into this bucket.
2. The namespace additions that you get by using exec.  _IoTHubModuelClient falls into this bucket.

It seems like exec() code should only use the second, and non-exec() code should use the first.  I'm saying this because I want to look at the strings that go into exec() and understand the patch completely.  I want to be able to copy the text from debug output and make a python file if I want.  Not that I'd ever do that, but it gives me another level to understand the code.  Another way to think about this is that patch.py is emitting code, and I want to be able to look at that code outside the context of patch.py and understand what it's doing.  In the debug output above, the first two lines make complete sense, but I can't understand what the third line is doing without knowing what method_object is set to.

Instead, I'm  wondering if the strings that we pass to exec would be better off looking like this:
```
# first two lines stay the same
INFO:azure.iot.device.iothub.patch: exec: from azure.iot.device import IoTHubModuleClient as _IoTHubModuleClient
INFO:azure.iot.device.iothub.patch: exec: def receive_input_message(self, input_name, block=True, timeout=None): return super(_IotHubModuleClient, self).receive_input_message(self=self, input_name=input_name, block=block, timeout=timeout))
# next two lines change
INFO:azure.iot.device.iothub.patch: exec: receive_input_message.__doc__ = _IoTHubModuleClient.receive_input_messages.__doc__
INFO:azure.iot.device.iothub.patch: exec: setattr(_IoTHubModuleClient, receive_input_messages, receive_input_message)
```

With this, I can look at debug output and see problems I might not otherwise see.

(oops, receive_input_messages isn't on GenericIoTHubClient.  My bad.)



---
In reply to: [315356054](https://github.com/Azure/azure-iot-sdk-python-preview/pull/164#discussion_r315356054) [](ancestors = 315356054)

Reply:
Another observation that comes out of this exercise is that your code leaves functions in the local namespace that don't need to be there.  in my example above, "receive_input_messages" is a function in the local namespace, and it doesn't need to be there.  No big deal, but the namespace exposed by the patch module is going to be a mish-mash of these dynamic functions from different classes and that just seems sloppy.

---
In reply to: [315386772](https://github.com/Azure/azure-iot-sdk-python-preview/pull/164#discussion_r315386772) [](ancestors = 315386772,315356054)

Reply:
Re: Comment 1 - I also was concerned about this, and made an effort to document and explain as much as I could with comments as a result. If you feel there's anything lacking, please let me know specifically, and I'll try to capture that inline.

---
In reply to: [315356054](https://github.com/Azure/azure-iot-sdk-python-preview/pull/164#discussion_r315356054) [](ancestors = 315356054)

Reply:
Re: Comment 2 - I think this is a great idea, both the logging, and the explicit division of namespacing. I'll add some additional logging now, but before I get further invested in this code by altering the namespacing, I want to make sure this works for the docs team first. Once we know they can ingest it properly, I'll come back and make these changes in a separate PR. I'll add an item to the taskboard for it now.

---
In reply to: [315453647](https://github.com/Azure/azure-iot-sdk-python-preview/pull/164#discussion_r315453647) [](ancestors = 315453647,315356054)

Reply:
Re: Comment 3 - It seems this problem would be resolved by manually maintaining a separate namespace for `exec`'d definitions, as mentioned in the above comment. They would still exist there, but it wouldn't be tied to the `patch.py` module.

---
In reply to: [315456629](https://github.com/Azure/azure-iot-sdk-python-preview/pull/164#discussion_r315456629) [](ancestors = 315456629,315453647,315356054)